tools/sysctlfix/main.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+	"github.com/spf13/pflag"
+)
+
+// This tool attempts to write a sysctl config file to the sysctl config directory with the highest precedence so
+// we can overwrite any other config and ensure correct sysctl options for Cilium to function.
+
+var (
+	flagSet = pflag.NewFlagSet(os.Args[0], pflag.ContinueOnError)
+
+	sysctlD = flagSet.String("sysctl-conf-dir", "/etc/sysctl.d/", "Path to the sysctl config directory")
+	// The 99-zzz prefix ensures our config file gets precedence over most if not all other files.
+	ciliumOverwrites = flagSet.String(
+		"sysctl-config-file",
+		"99-zzz-override_cilium.conf",
+		"Filename of the cilium sysctl overwrites config file",
+	)
+	// Name of the systemd-sysctl unit to restart after making changes
+	sysctlUnit = flagSet.String(
+		"systemd-sysctl-unit",
+		"systemd-sysctl.service",
+		"Name of the systemd sysctl unit to reload",
+	)
+)
+
+var sysctlConfig = strings.Join([]string{
+	"# Disable rp_filter on Cilium interfaces since it may cause mangled packets to be dropped",
+	"net.ipv4.conf.all.rp_filter = 0",
```

Comment:
Aren't we changing the rp_filter setting of other non-Cilium ifaces? If yes, then this might be undesirable behavior by users who are concerned about the packet spoofing.

Reply:
Yes, it does. According to @rastislavs this is required to fix #19761. Perhaps we can find a middle ground between disabling on all interfaces and just the Cilium interfaces? 

If `rp_filter` on host interfaces really does create connectivity problems, perhaps we should disable it and fast-track spoofing detection in Cilium, so we can guarantee it is compatible?

Reply:
> According to @rastislavs this is required to fix

I think it makes sense to first understand which packet path was broken due to rp_filter=2 in that case.

Reply:
That seems fair, I will revert this PR to only setting `rp_filter` for Cilium interfaces so we can get this PR merged for the other issues. Then we will do a followup PR once we have more details regarding #19761.

Reply:
But isn't Cilium already setting `net.ipv4.conf.all.rp_filter` to `0` upon its startup ([here](https://github.com/cilium/cilium/blob/v1.11.5/pkg/datapath/loader/base.go#L245))? At least this is a log message that I can see in Cilium logs when it starts:
```
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.all.rp_filter sysParamValue=0
```
So why should this be any different to that logic?

Reply:
We set the config to the `Loose` mode while installing ENI rules on EKS clusters - http://github.com/cilium/cilium/blob/master/pkg/datapath/loader/base.go#L146-L146. We should ideally have a consistent way of setting the config. I would suggest investigating this further. You may refer to this issue https://github.com/cilium/cilium/issues/13130 to get pointers for debugging.

Reply:
> But isn't Cilium already setting net.ipv4.conf.all.rp_filter to 0 upon its startup

Yes, but this only changes the runtime option, if systemd is triggered for any reason it can just change the runtime setting to match the config. By setting the desired sysctl parameter in systemd config, we ensure that the value stays even after systemd config reloads

Reply:
Please see my latest comment on https://github.com/cilium/cilium/issues/19761#issuecomment-1150751732

Reply:
https://github.com/cilium/cilium/pull/20072#issuecomment-1156495539